In [1]:
!wget --no-check-certificate \
  https://s3-ap-southeast-1.amazonaws.com/he-public-data/dataset52a7b21.zip \
  -O /tmp/product_browse_node.zip
  

--2021-07-30 18:47:32--  https://s3-ap-southeast-1.amazonaws.com/he-public-data/dataset52a7b21.zip
Resolving s3-ap-southeast-1.amazonaws.com (s3-ap-southeast-1.amazonaws.com)... 52.219.36.74
Connecting to s3-ap-southeast-1.amazonaws.com (s3-ap-southeast-1.amazonaws.com)|52.219.36.74|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1061576029 (1012M) [binary/octet-stream]
Saving to: ‘/tmp/product_browse_node.zip’

/tmp/product_browse 100%[===================>]   1012M  17.5MB/s    in 60s     

2021-07-30 18:48:33 (16.8 MB/s) - ‘/tmp/product_browse_node.zip’ saved [1061576029/1061576029]



In [2]:
import os
import zipfile

local_zip = '/tmp/product_browse_node.zip'

zip_ref = zipfile.ZipFile(local_zip, 'r')

zip_ref.extractall('/tmp')
zip_ref.close()

In [3]:
import pandas as pd
import tensorflow as tf
import numpy as np

df1 = pd.read_csv("/tmp/dataset/train.csv", quoting=3, escapechar="\\")
df1



TITLE  ... BROWSE_NODE_ID
0               Pete The Cat Bedtime Blues Doll, 14.5 Inch  ...              0
1        The New Yorker NYHM014 Refrigerator Magnet, 2 ...  ...              1
2        The Ultimate Self-Sufficiency Handbook: A Comp...  ...              2
3         Amway Nutrilite Kids Chewable Iron Tablets (100)  ...              3
4        Teacher Planner Company A4 6 Lesson Academic T...  ...              4
...                                                    ...  ...            ...
2903019  Premium Aviator Sunglasses - HD Polarized (Bri...  ...           1040
2903020  Social Distance Stickers - Set of 5 Sticker Sl...  ...          15199
2903021  Torr-to Face Shield PACK OF 5 with Adjustable ...  ...        1044933
2903022  Type-C to 3.5 MM for Oppo R17 Pro Type-C to 3....  ...          14790
2903023                        KNG Orange Polo Tshirt (38)  ...           1213

[2903024 rows x 5 columns]

In [4]:
df2 = pd.read_csv("/tmp/dataset/test.csv", quoting=3, escapechar="\\")
df2


PRODUCT_ID  ...           BRAND
0                1  ...         Command
1                2  ...          O'Neal
2                3  ...  Boelter Brands
3                4  ...       Panasonic
4                5  ...            Zero
...            ...  ...             ...
110770      110771  ...         Generic
110771      110772  ...     Grin Health
110772      110773  ...           ASIAN
110773      110774  ...     Intermarket
110774      110775  ...         Artifie

[110775 rows x 5 columns]

In [5]:
pd.set_option('display.max_colwidth',None)
df1.head(50)

TITLE  ... BROWSE_NODE_ID
0                                                                                                                Pete The Cat Bedtime Blues Doll, 14.5 Inch  ...              0
1                                                                                                       The New Yorker NYHM014 Refrigerator Magnet, 2 x 3.5  ...              1
2                  The Ultimate Self-Sufficiency Handbook: A Complete Guide to Baking, Crafts, Gardening, Preserving Your Harvest, Raising Animals and More  ...              2
3                                                                                                          Amway Nutrilite Kids Chewable Iron Tablets (100)  ...              3
4                                                                           Teacher Planner Company A4 6 Lesson Academic Teacher Planner - Leatherette Pink  ...              4
5                                                                                               Men'S Full Sleeve Raglan T-Shirts Denim T-Shirt M Size Only  ...              5
6                                                                                                                     Glance Women's Wallet (Black) (LW-21)  ...              6
7                                            Wild Animals Hungry Brain Educational Flash Cards for Enhancing Early Learning of Your Kids(3 Month to 6 Year)  ...              7
8                                                Hindi Barakhadi Educational Flash Cards for Kids Enhance Early Learning of Your Kids (3 Months to 6 Years)  ...              7
9                                                                                    Chapin 1380 Premier Pro 3-Gallon Tri-Poxy Steel Sprayer for Fertilizer  ...              9
10                                                                                        Slo-Niacin Polygel Controlled-Release Niacin, 750 mg, 100 Tablets  ...             10
11                                                                                               White : Braun KF400-WH Aromaster 10-Cup Coffeemaker, White  ...             11
12                                                                       Yamaha PCMSKIT Portable Electronic Keyboard Adapter with PC Joystick to MIDI Cable  ...             12
13                                                                                                                         Microsoft Sidewinder Gamepad Pro  ...             13
14                                                                                                      Jiffy Steamer with Plastic Steam Head (Beige, 120V)  ...             14
15                                                                                             YASHICA Open Box OPTICAL OPTICAL Zoomate 140 35mm Camera Kit  ...             15
16                                                                                                OXO Good Grips BPA-Free Turkey Baster with Cleaning Brush  ...             16
17                                                                                                    Sony VAIO RoomLink Network Media Receiver (PCNA-MR10)  ...             17
18                                                                                Westcott 5684 5 x 6 Feet Collapsible Backdrop (Costa Brava Blue/Canberra)  ...             18
19                                                                                                InterDesign 14 by 9 by 5-Inch Basic Basket, Medium, White  ...             19
20                                                                                                         Woodard & Charles 6.5-Inch Individual Salad Bowl  ...             20
21                                                                                                                       Meco Swinger Charcoal Grill, Black  ...             21
22                                                                                                    

In [6]:
print(df1.shape)
df2.shape

(2903024, 5)


(110775, 5)

#data analysis
- count everything the data

In [7]:
t=df1["BROWSE_NODE_ID"].max()
df1[2893570:2893575]

TITLE  ... BROWSE_NODE_ID
2893570                                                                                                              AIRCON Hand Dryer  ...           3172
2893571  YOYOSHome Luminous Japanese Anime Cosplay Daypack Bookbag Laptop Bag Backpack School Bag with USB Charging Port (Tokyo Ghoul)  ...           1339
2893572                                         Clarks Men's Polyflex Mule Brown Leather Sandals-6.5 UK/India (40 EU) (91261281407065)  ...            748
2893573                                                  Toddler Boy 4T 5T Red With Cats Clip On Cotton Bow Tie Bowtie by amy2004marie  ...            856
2893574                                                                           Yonex Junior Polo T-Shirt 932-28 (Strong Blue, J160)  ...        2893574

[5 rows x 5 columns]

In [8]:
df1.sort_values(by='BROWSE_NODE_ID').head(100)

TITLE  ... BROWSE_NODE_ID
0                                                                                                                                                       Pete The Cat Bedtime Blues Doll, 14.5 Inch  ...              0
1773215                                                                                                                                                   Barbie Chelsea Assortment 2, Multi Color  ...              0
1694581                                                                                                   ELECTROPRIME Realistic Vinyl Baby Girl Doll Lifelike Singing Walking Doll Toy Child Gift  ...              0
2294866                                                                                                                     Manhattan Toy Groovy Girls Style Scents Lilly and Parfait Fashion Doll  ...              0
1469000                                                                                                               ZWSISU 7 Outfits American Girl Doll Clothes Accessories (18 Inch) - Set of 9  ...              0
...                                                                                                                                                                                            ...  ...            ...
2389064                                                                                                        AKD06 Direct 14pcs Kids Portable Toolbox Toy Simulation Disassembly Repair Tool Set  ...              0
1024479  Brand Conquer Dancing Fairy Princes Angel Girl Robot with Lights and Music/ Angel Girl Dancing Doll Robot Music Baby Toy 360 Rotational 3D LED Lights for Birthday Gift for Kids Children  ...              0
1467793                                                                                                                                   BARBIETM The Pearl Princess 2-in-1 Mermaid Princess Doll  ...              0
306983                                                                                                                   Tickles Blue Laughing Baby Stuffed Soft Plush Toy Love Girl 36 cm AT-T680  ...              0
2614794                                                                                                                                              Hearts for Hearts Girls Nahji from India Doll  ...              0

[100 rows x 5 columns]

In [9]:
df1.isnull().sum()

TITLE                 71
DESCRIPTION       723664
BULLET_POINTS     166263
BRAND              56737
BROWSE_NODE_ID         0
dtype: int64

##Data preprocessing :
- dividing datasets into three parts train, test(already given), cross-validation(pick from training set)
- randomize the dataset before dividing 

##Data cleaning :
- duplicates
- NaN
- html tags
- punctuation
- []
- cannot remove , in bullet point

In [10]:
import nltk
from nltk.tokenize import sent_tokenize,word_tokenize
nltk.download('punkt')


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [11]:
train1 =  df1.copy(deep=True)
print(train1.isna().sum())
train1.fillna("", inplace = True)
# train1.drop('BROWSE_NODE_ID',axis=1,inplace=True)
train1

TITLE                 71
DESCRIPTION       723664
BULLET_POINTS     166263
BRAND              56737
BROWSE_NODE_ID         0
dtype: int64


TITLE  ... BROWSE_NODE_ID
0                                                                                                                                                 Pete The Cat Bedtime Blues Doll, 14.5 Inch  ...              0
1                                                                                                                                        The New Yorker NYHM014 Refrigerator Magnet, 2 x 3.5  ...              1
2                                                   The Ultimate Self-Sufficiency Handbook: A Complete Guide to Baking, Crafts, Gardening, Preserving Your Harvest, Raising Animals and More  ...              2
3                                                                                                                                           Amway Nutrilite Kids Chewable Iron Tablets (100)  ...              3
4                                                                                                            Teacher Planner Company A4 6 Lesson Academic Teacher Planner - Leatherette Pink  ...              4
...                                                                                                                                                                                      ...  ...            ...
2903019                                                                                                                  Premium Aviator Sunglasses - HD Polarized (Bright Gold/Ocean Green)  ...           1040
2903020                                                                            Social Distance Stickers - Set of 5 Sticker Slip Resistance, Durable & Strong Adhesive, Easy to Paste-R/Y  ...          15199
2903021  Torr-to Face Shield PACK OF 5 with Adjustable Elastic Strap Anti-Splash Protective Facial Transparent Cover Full Face Visor with Eye & Head Protection for Men & Women, 350 Micron,  ...        1044933
2903022                                                      Type-C to 3.5 MM for Oppo R17 Pro Type-C to 3.5 MM Audio Jack Adapter Converter for CDLA Type, USB C Connector Splitter (White)  ...          14790
2903023                                                                                                                                                          KNG Orange Polo Tshirt (38)  ...           1213

[2903024 rows x 5 columns]

In [ ]:
c1 =train1.TITLE
c2 =train1.DESCRIPTION
c3 =train1.BULLET_POINTS
# word_tokenize(train1.DESCRIPTION[1])
print(c1)
print(c2)
print(c3)

0                                                                                                                                                   Pete The Cat Bedtime Blues Doll, 14.5 Inch
1                                                                                                                                          The New Yorker NYHM014 Refrigerator Magnet, 2 x 3.5
2                                                     The Ultimate Self-Sufficiency Handbook: A Complete Guide to Baking, Crafts, Gardening, Preserving Your Harvest, Raising Animals and More
3                                                                                                                                             Amway Nutrilite Kids Chewable Iron Tablets (100)
4                                                                                                              Teacher Planner Company A4 6 Lesson Academic Teacher Planner - Leatherette Pink
                                             

In [12]:
import re
from nltk.corpus import stopwords
nltk.download('stopwords')
import string

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [13]:
def cleaning(text):
    text = str(text).lower()
    text = re.sub('\[.*?\]', '', text)
    text = re.sub('https?://\S+|www\.\S+', '', text)
    text = re.sub('<.*?>+', '', text)
    text = re.sub('[%s]' % re.escape(string.punctuation), '', text)
    text = re.sub('\n', '', text)
    text = re.sub('\w*\d\w*', '', text)

    return text

creating a small set of data

In [14]:
t1 = (train1[:20000].copy(deep=True))
t1

TITLE  ... BROWSE_NODE_ID
0                                                                                                    Pete The Cat Bedtime Blues Doll, 14.5 Inch  ...              0
1                                                                                           The New Yorker NYHM014 Refrigerator Magnet, 2 x 3.5  ...              1
2      The Ultimate Self-Sufficiency Handbook: A Complete Guide to Baking, Crafts, Gardening, Preserving Your Harvest, Raising Animals and More  ...              2
3                                                                                              Amway Nutrilite Kids Chewable Iron Tablets (100)  ...              3
4                                                               Teacher Planner Company A4 6 Lesson Academic Teacher Planner - Leatherette Pink  ...              4
...                                                                                                                                         ...  ...            ...
19995                                                                                                      Dixcy Scott Women's Skinny Fit Capri  ...           1058
19996                 ELECTROPRIME 1U Sheilded Cable Management 19 Universal Keystone Jacks Multimedia Panel 24-Port Rackmount for Data Cabinet  ...           2205
19997                                 ELECTROPRIME 4pcs Universal Silver Tone Chrome Plastic Auto Car Wheel Center Hub Cap Cover Dust Againster  ...           2161
19998                                                ELECTROPRIME 2 Pcs 4.5cm Dia Mesh Car Tweeter Trumpet Dustproof Cover for 2014 Toyota RAV4  ...          19998
19999                                                                Reebok Men's Twist Run Multicolor Shoes- 8 UK/India (42 EU)(9 US) (CN4451)  ...            701

[20000 rows x 5 columns]

In [15]:
t1['TITLE'] = t1['TITLE'].apply(cleaning)
t1

TITLE  ... BROWSE_NODE_ID
0                                                                                                   pete the cat bedtime blues doll  inch  ...              0
1                                                                                                 the new yorker  refrigerator magnet  x   ...              1
2      the ultimate selfsufficiency handbook a complete guide to baking crafts gardening preserving your harvest raising animals and more  ...              2
3                                                                                             amway nutrilite kids chewable iron tablets   ...              3
4                                                             teacher planner company   lesson academic teacher planner  leatherette pink  ...              4
...                                                                                                                                   ...  ...            ...
19995                                                                                                 dixcy scott womens skinny fit capri  ...           1058
19996                      electroprime  sheilded cable management  universal keystone jacks multimedia panel  rackmount for data cabinet  ...           2205
19997                               electroprime  universal silver tone chrome plastic auto car wheel center hub cap cover dust againster  ...           2161
19998                                                        electroprime  pcs  dia mesh car tweeter trumpet dustproof cover for  toyota   ...          19998
19999                                                                               reebok mens twist run multicolor shoes  ukindia   us   ...            701

[20000 rows x 5 columns]

In [16]:
# Removing stop words
stop_words = stopwords.words('english')
more = ['mm', 'cm', 'x', 'inch', 'inches', 'millimeter', 'centimeter']
stop_words = stop_words + more
def sw_rem(text):
    text = ' '.join(word for word in text.split(' ') if word not in stop_words)
    return text


t1['TITLE'] = t1['TITLE'].apply(sw_rem)
t1['TITLE']
t1

TITLE  ... BROWSE_NODE_ID
0                                                                                     pete cat bedtime blues doll   ...              0
1                                                                                new yorker  refrigerator magnet    ...              1
2      ultimate selfsufficiency handbook complete guide baking crafts gardening preserving harvest raising animals  ...              2
3                                                                      amway nutrilite kids chewable iron tablets   ...              3
4                                      teacher planner company   lesson academic teacher planner  leatherette pink  ...              4
...                                                                                                            ...  ...            ...
19995                                                                          dixcy scott womens skinny fit capri  ...           1058
19996   electroprime  sheilded cable management  universal keystone jacks multimedia panel  rackmount data cabinet  ...           2205
19997        electroprime  universal silver tone chrome plastic auto car wheel center hub cap cover dust againster  ...           2161
19998                                     electroprime  pcs  dia mesh car tweeter trumpet dustproof cover  toyota   ...          19998
19999                                                        reebok mens twist run multicolor shoes  ukindia   us   ...            701

[20000 rows x 5 columns]

In [18]:
t1['DESCRIPTION'] = t1['DESCRIPTION'].apply(cleaning)
t1

TITLE  ... BROWSE_NODE_ID
0                                                                                     pete cat bedtime blues doll   ...              0
1                                                                                new yorker  refrigerator magnet    ...              1
2      ultimate selfsufficiency handbook complete guide baking crafts gardening preserving harvest raising animals  ...              2
3                                                                      amway nutrilite kids chewable iron tablets   ...              3
4                                      teacher planner company   lesson academic teacher planner  leatherette pink  ...              4
...                                                                                                            ...  ...            ...
19995                                                                          dixcy scott womens skinny fit capri  ...           1058
19996   electroprime  sheilded cable management  universal keystone jacks multimedia panel  rackmount data cabinet  ...           2205
19997        electroprime  universal silver tone chrome plastic auto car wheel center hub cap cover dust againster  ...           2161
19998                                     electroprime  pcs  dia mesh car tweeter trumpet dustproof cover  toyota   ...          19998
19999                                                        reebok mens twist run multicolor shoes  ukindia   us   ...            701

[20000 rows x 5 columns]

In [19]:
t1['DESCRIPTION'] = t1['DESCRIPTION'].apply(sw_rem)
t1

TITLE  ... BROWSE_NODE_ID
0                                                                                     pete cat bedtime blues doll   ...              0
1                                                                                new yorker  refrigerator magnet    ...              1
2      ultimate selfsufficiency handbook complete guide baking crafts gardening preserving harvest raising animals  ...              2
3                                                                      amway nutrilite kids chewable iron tablets   ...              3
4                                      teacher planner company   lesson academic teacher planner  leatherette pink  ...              4
...                                                                                                            ...  ...            ...
19995                                                                          dixcy scott womens skinny fit capri  ...           1058
19996   electroprime  sheilded cable management  universal keystone jacks multimedia panel  rackmount data cabinet  ...           2205
19997        electroprime  universal silver tone chrome plastic auto car wheel center hub cap cover dust againster  ...           2161
19998                                     electroprime  pcs  dia mesh car tweeter trumpet dustproof cover  toyota   ...          19998
19999                                                        reebok mens twist run multicolor shoes  ukindia   us   ...            701

[20000 rows x 5 columns]

In [20]:
# # train1.BULLET_POINTS = df1.BULLET_POINTS
t1['BULLET_POINTS']
test_df=t1.copy(deep=True)
test_df

# # def special_cleaning(text):
# #   text = str(text).lower()
# #   # text = re.sub('\[.*?\]', '', text)
# #   text = re.sub('https?://\S+|www\.\S+', '', text)
# #   text = re.sub('<.*?>+', '', text)
# #   text = re.sub('[%s]' % re.escape(string.punctuation), '', text)
# #   text = re.sub('\n', '', text)
# #   text = re.sub('\w*\d\w*', '', text)
# #   return text

test_df['BULLET_POINTS']= test_df['BULLET_POINTS'].apply(sw_rem)
test_df





TITLE  ... BROWSE_NODE_ID
0                                                                                     pete cat bedtime blues doll   ...              0
1                                                                                new yorker  refrigerator magnet    ...              1
2      ultimate selfsufficiency handbook complete guide baking crafts gardening preserving harvest raising animals  ...              2
3                                                                      amway nutrilite kids chewable iron tablets   ...              3
4                                      teacher planner company   lesson academic teacher planner  leatherette pink  ...              4
...                                                                                                            ...  ...            ...
19995                                                                          dixcy scott womens skinny fit capri  ...           1058
19996   electroprime  sheilded cable management  universal keystone jacks multimedia panel  rackmount data cabinet  ...           2205
19997        electroprime  universal silver tone chrome plastic auto car wheel center hub cap cover dust againster  ...           2161
19998                                     electroprime  pcs  dia mesh car tweeter trumpet dustproof cover  toyota   ...          19998
19999                                                        reebok mens twist run multicolor shoes  ukindia   us   ...            701

[20000 rows x 5 columns]

In [21]:
# train1
REPLACE_BY_SPACE_RE = re.compile('[/(){}\[\]\|@,;]')
BAD_SYMBOLS_RE = re.compile('[^0-9a-z #+_]')
STOPWORDS = set(stopwords.words('english'))

def clean_text(text):
    text = text.lower() # lowercase text
    text = REPLACE_BY_SPACE_RE.sub(' ', text) # replace REPLACE_BY_SPACE_RE symbols by space in text. substitute the matched string in REPLACE_BY_SPACE_RE with space.
    text = BAD_SYMBOLS_RE.sub('', text) # remove symbols which are in BAD_SYMBOLS_RE from text. substitute the matched string in BAD_SYMBOLS_RE with nothing. 
    text = text.replace('x', '')
#    text = re.sub(r'\W+', '', text)
    # text = ' '.join(word for word in text.split() if word not in STOPWORDS) # remove stopwors from text
    return text
test_df['BULLET_POINTS'] = test_df['BULLET_POINTS'].apply(clean_text)
test_df['BULLET_POINTS'] = test_df['BULLET_POINTS'].str.replace('\d+', '')
test_df

TITLE  ... BROWSE_NODE_ID
0                                                                                     pete cat bedtime blues doll   ...              0
1                                                                                new yorker  refrigerator magnet    ...              1
2      ultimate selfsufficiency handbook complete guide baking crafts gardening preserving harvest raising animals  ...              2
3                                                                      amway nutrilite kids chewable iron tablets   ...              3
4                                      teacher planner company   lesson academic teacher planner  leatherette pink  ...              4
...                                                                                                            ...  ...            ...
19995                                                                          dixcy scott womens skinny fit capri  ...           1058
19996   electroprime  sheilded cable management  universal keystone jacks multimedia panel  rackmount data cabinet  ...           2205
19997        electroprime  universal silver tone chrome plastic auto car wheel center hub cap cover dust againster  ...           2161
19998                                     electroprime  pcs  dia mesh car tweeter trumpet dustproof cover  toyota   ...          19998
19999                                                        reebok mens twist run multicolor shoes  ukindia   us   ...            701

[20000 rows x 5 columns]

In [22]:
test_df


TITLE  ... BROWSE_NODE_ID
0                                                                                     pete cat bedtime blues doll   ...              0
1                                                                                new yorker  refrigerator magnet    ...              1
2      ultimate selfsufficiency handbook complete guide baking crafts gardening preserving harvest raising animals  ...              2
3                                                                      amway nutrilite kids chewable iron tablets   ...              3
4                                      teacher planner company   lesson academic teacher planner  leatherette pink  ...              4
...                                                                                                            ...  ...            ...
19995                                                                          dixcy scott womens skinny fit capri  ...           1058
19996   electroprime  sheilded cable management  universal keystone jacks multimedia panel  rackmount data cabinet  ...           2205
19997        electroprime  universal silver tone chrome plastic auto car wheel center hub cap cover dust againster  ...           2161
19998                                     electroprime  pcs  dia mesh car tweeter trumpet dustproof cover  toyota   ...          19998
19999                                                        reebok mens twist run multicolor shoes  ukindia   us   ...            701

[20000 rows x 5 columns]

In [21]:
train1

TITLE  ... BROWSE_NODE_ID
0                                                                                                                                                 Pete The Cat Bedtime Blues Doll, 14.5 Inch  ...              0
1                                                                                                                                        The New Yorker NYHM014 Refrigerator Magnet, 2 x 3.5  ...              1
2                                                   The Ultimate Self-Sufficiency Handbook: A Complete Guide to Baking, Crafts, Gardening, Preserving Your Harvest, Raising Animals and More  ...              2
3                                                                                                                                           Amway Nutrilite Kids Chewable Iron Tablets (100)  ...              3
4                                                                                                            Teacher Planner Company A4 6 Lesson Academic Teacher Planner - Leatherette Pink  ...              4
...                                                                                                                                                                                      ...  ...            ...
2903019                                                                                                                  Premium Aviator Sunglasses - HD Polarized (Bright Gold/Ocean Green)  ...           1040
2903020                                                                            Social Distance Stickers - Set of 5 Sticker Slip Resistance, Durable & Strong Adhesive, Easy to Paste-R/Y  ...          15199
2903021  Torr-to Face Shield PACK OF 5 with Adjustable Elastic Strap Anti-Splash Protective Facial Transparent Cover Full Face Visor with Eye & Head Protection for Men & Women, 350 Micron,  ...        1044933
2903022                                                      Type-C to 3.5 MM for Oppo R17 Pro Type-C to 3.5 MM Audio Jack Adapter Converter for CDLA Type, USB C Connector Splitter (White)  ...          14790
2903023                                                                                                                                                          KNG Orange Polo Tshirt (38)  ...           1213

[2903024 rows x 5 columns]

In [22]:
t1

TITLE  ... BROWSE_NODE_ID
0                                                                                     pete cat bedtime blues doll   ...              0
1                                                                                new yorker  refrigerator magnet    ...              1
2      ultimate selfsufficiency handbook complete guide baking crafts gardening preserving harvest raising animals  ...              2
3                                                                      amway nutrilite kids chewable iron tablets   ...              3
4                                      teacher planner company   lesson academic teacher planner  leatherette pink  ...              4
...                                                                                                            ...  ...            ...
19995                                                                          dixcy scott womens skinny fit capri  ...           1058
19996   electroprime  sheilded cable management  universal keystone jacks multimedia panel  rackmount data cabinet  ...           2205
19997        electroprime  universal silver tone chrome plastic auto car wheel center hub cap cover dust againster  ...           2161
19998                                     electroprime  pcs  dia mesh car tweeter trumpet dustproof cover  toyota   ...          19998
19999                                                        reebok mens twist run multicolor shoes  ukindia   us   ...            701

[20000 rows x 5 columns]

#Stemming and lemmatization
Documents and other forms of text use different forms of the same words, such as play, playing, played. There are families of derivationally related words that have similar meanings. Our main task with stemming and lemmatization is to reduce all these derived words into the parent/family word, therefore reducing the total vocabulary while retaining information.

##Stemming 
- Omits the ends of words to achieve the goal correctly, this works most of the times and can also remove the derivational suffix

##Lemmatization - 
- Working with a vocabulary and morphological analysis of wrods, removing inflectional endings only and returning the base and dictionary form of a word.
- As we do not require much emphasis on words, we will focus more on stemming than lemmatization,.

##Stemming algorithms
- We have multiple algorithms to achieve our stemming goals, some of them are as follows:

####PorterStemmer 
- Fast and efficient. Strips off the end (suffix) to produce the stems. It does not follow linguistics but rather a set of 05 rules for diferent cases.

####SnowballStemmer 
- Generate a set of rules for any language. These are useful for non-english stemming tasks.

####LancasterStemmer 
- Iterative algorithm, uses about 120 rules, it tries to find an applicable rule by the last character of each word. The last character may be omitted or replaced.

In [23]:
test_df2 = test_df.copy(deep=True)
test_df2
def stopwords2(text):
  text = ' '.join(word for word in text.split() if word not in stop_words) # remove stopwors from text
  return text

test_df2['BULLET_POINTS']=test_df2['BULLET_POINTS'].apply(stopwords2)


In [24]:
test_df2

TITLE  ... BROWSE_NODE_ID
0                                                                                     pete cat bedtime blues doll   ...              0
1                                                                                new yorker  refrigerator magnet    ...              1
2      ultimate selfsufficiency handbook complete guide baking crafts gardening preserving harvest raising animals  ...              2
3                                                                      amway nutrilite kids chewable iron tablets   ...              3
4                                      teacher planner company   lesson academic teacher planner  leatherette pink  ...              4
...                                                                                                            ...  ...            ...
19995                                                                          dixcy scott womens skinny fit capri  ...           1058
19996   electroprime  sheilded cable management  universal keystone jacks multimedia panel  rackmount data cabinet  ...           2205
19997        electroprime  universal silver tone chrome plastic auto car wheel center hub cap cover dust againster  ...           2161
19998                                     electroprime  pcs  dia mesh car tweeter trumpet dustproof cover  toyota   ...          19998
19999                                                        reebok mens twist run multicolor shoes  ukindia   us   ...            701

[20000 rows x 5 columns]

In [25]:
from nltk import stem
from nltk.corpus import treebank
stems = nltk.PorterStemmer()

def stemming(text):
    text = ' '.join(stems.stem(word) for word in text.split())
    return text


In [26]:
test_df3 = test_df2.copy(deep=True)
test_df3

TITLE  ... BROWSE_NODE_ID
0                                                                                     pete cat bedtime blues doll   ...              0
1                                                                                new yorker  refrigerator magnet    ...              1
2      ultimate selfsufficiency handbook complete guide baking crafts gardening preserving harvest raising animals  ...              2
3                                                                      amway nutrilite kids chewable iron tablets   ...              3
4                                      teacher planner company   lesson academic teacher planner  leatherette pink  ...              4
...                                                                                                            ...  ...            ...
19995                                                                          dixcy scott womens skinny fit capri  ...           1058
19996   electroprime  sheilded cable management  universal keystone jacks multimedia panel  rackmount data cabinet  ...           2205
19997        electroprime  universal silver tone chrome plastic auto car wheel center hub cap cover dust againster  ...           2161
19998                                     electroprime  pcs  dia mesh car tweeter trumpet dustproof cover  toyota   ...          19998
19999                                                        reebok mens twist run multicolor shoes  ukindia   us   ...            701

[20000 rows x 5 columns]

In [27]:
test_df3['DESCRIPTION'] = test_df3['DESCRIPTION'].apply(stemming)
test_df3

TITLE  ... BROWSE_NODE_ID
0                                                                                     pete cat bedtime blues doll   ...              0
1                                                                                new yorker  refrigerator magnet    ...              1
2      ultimate selfsufficiency handbook complete guide baking crafts gardening preserving harvest raising animals  ...              2
3                                                                      amway nutrilite kids chewable iron tablets   ...              3
4                                      teacher planner company   lesson academic teacher planner  leatherette pink  ...              4
...                                                                                                            ...  ...            ...
19995                                                                          dixcy scott womens skinny fit capri  ...           1058
19996   electroprime  sheilded cable management  universal keystone jacks multimedia panel  rackmount data cabinet  ...           2205
19997        electroprime  universal silver tone chrome plastic auto car wheel center hub cap cover dust againster  ...           2161
19998                                     electroprime  pcs  dia mesh car tweeter trumpet dustproof cover  toyota   ...          19998
19999                                                        reebok mens twist run multicolor shoes  ukindia   us   ...            701

[20000 rows x 5 columns]

In [28]:
test_df3['BULLET_POINTS'] = test_df3['BULLET_POINTS'].apply(stemming)
test_df3

TITLE  ... BROWSE_NODE_ID
0                                                                                     pete cat bedtime blues doll   ...              0
1                                                                                new yorker  refrigerator magnet    ...              1
2      ultimate selfsufficiency handbook complete guide baking crafts gardening preserving harvest raising animals  ...              2
3                                                                      amway nutrilite kids chewable iron tablets   ...              3
4                                      teacher planner company   lesson academic teacher planner  leatherette pink  ...              4
...                                                                                                            ...  ...            ...
19995                                                                          dixcy scott womens skinny fit capri  ...           1058
19996   electroprime  sheilded cable management  universal keystone jacks multimedia panel  rackmount data cabinet  ...           2205
19997        electroprime  universal silver tone chrome plastic auto car wheel center hub cap cover dust againster  ...           2161
19998                                     electroprime  pcs  dia mesh car tweeter trumpet dustproof cover  toyota   ...          19998
19999                                                        reebok mens twist run multicolor shoes  ukindia   us   ...            701

[20000 rows x 5 columns]

In [ ]:
# sample_train= train1.loc[:20000]
# sample_train

In [29]:
cross_val= (test_df3.copy(deep=True)).sample(frac=.2)
cross_val

TITLE  ... BROWSE_NODE_ID
11825                                                              farmoura black flower glass corner shelf brass brackets  size   transparent  ...           5484
19713                                                                                    kiddy depot –hindi consonants  alphabet tracing board  ...            723
14386                                                         silver shine medical examination disposable hand  latex examination gloves pack   ...           3503
1404                                                                                             monoprice    inceiling speaker ribbon tweeter  ...            660
19854                                                                                                     ca brown floral print lingerie thong  ...           7917
...                                                                                                                                        ...  ...            ...
14007                                                         ringjewel  sterling silver heart cut near white moissanite solitaire ring size    ...             22
9484                                                                                           ssc mens ultimate polyster sports tshirt medium  ...              5
44     oster  pc kit containing osters stiff grooming brush soft finishing brush  coarse curry comb  mane tail brush  mane tail comb hoof pick  ...             44
18405                                                                           excalibur unlimited mens plain regular fit formal shirt  pack   ...           1254
1204                                                   creative  portable bluetooth audio transceiver aptx low latency pc mac  nintendo switch  ...            487

[4000 rows x 5 columns]

First tokenization
- count unique words


In [30]:
test_df4 = test_df3.copy(deep=True)
test_df4

TITLE  ... BROWSE_NODE_ID
0                                                                                     pete cat bedtime blues doll   ...              0
1                                                                                new yorker  refrigerator magnet    ...              1
2      ultimate selfsufficiency handbook complete guide baking crafts gardening preserving harvest raising animals  ...              2
3                                                                      amway nutrilite kids chewable iron tablets   ...              3
4                                      teacher planner company   lesson academic teacher planner  leatherette pink  ...              4
...                                                                                                            ...  ...            ...
19995                                                                          dixcy scott womens skinny fit capri  ...           1058
19996   electroprime  sheilded cable management  universal keystone jacks multimedia panel  rackmount data cabinet  ...           2205
19997        electroprime  universal silver tone chrome plastic auto car wheel center hub cap cover dust againster  ...           2161
19998                                     electroprime  pcs  dia mesh car tweeter trumpet dustproof cover  toyota   ...          19998
19999                                                        reebok mens twist run multicolor shoes  ukindia   us   ...            701

[20000 rows x 5 columns]

In [31]:
from keras.preprocessing.text import Tokenizer

In [32]:
test_df4=test_df4.drop(columns=['BROWSE_NODE_ID'])
test_df4

TITLE  ...           BRAND
0                                                                                     pete cat bedtime blues doll   ...     MerryMakers
1                                                                                new yorker  refrigerator magnet    ...  The New Yorker
2      ultimate selfsufficiency handbook complete guide baking crafts gardening preserving harvest raising animals  ...          imusti
3                                                                      amway nutrilite kids chewable iron tablets   ...           Amway
4                                      teacher planner company   lesson academic teacher planner  leatherette pink  ...                
...                                                                                                            ...  ...             ...
19995                                                                          dixcy scott womens skinny fit capri  ...     Dixcy Scott
19996   electroprime  sheilded cable management  universal keystone jacks multimedia panel  rackmount data cabinet  ...    ELECTROPRIME
19997        electroprime  universal silver tone chrome plastic auto car wheel center hub cap cover dust againster  ...    ELECTROPRIME
19998                                     electroprime  pcs  dia mesh car tweeter trumpet dustproof cover  toyota   ...    ELECTROPRIME
19999                                                        reebok mens twist run multicolor shoes  ukindia   us   ...          Reebok

[20000 rows x 4 columns]

In [ ]:
import re
import string
import numpy as np 
import random
import pandas as pd 
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
from plotly import graph_objs as go
import plotly.express as px
import plotly.figure_factory as ff
from collections import Counter

from PIL import Image
from wordcloud import WordCloud, STOPWORDS, ImageColorGenerator


import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize

from tqdm import tqdm
import os
import nltk
import spacy
import random
from spacy.util import compounding
from spacy.util import minibatch

from collections import defaultdict
from collections import Counter

import keras
from keras.models import Sequential
from keras.initializers import Constant
from keras.layers import LSTM, Embedding,BatchNormalization, Dense, TimeDistributed, Dropout, Bidirectional, Flatten, GlobalMaxPool1D
from keras.preprocessing.text import Tokenizer
from keras.preprocessing.sequence import pad_sequences
from keras.layers.embeddings import Embedding
from keras.callbacks import ModelCheckpoint, ReduceLROnPlateau
from keras.optimizers import Adam

from sklearn.metrics import precision_score, recall_score, f1_score, classification_report,accuracy_score

In [33]:
np.max(test_df4['DESCRIPTION'].apply(lambda x: len(x.split())).values)
# Finding maximum length of text message

270

i should merge all the text in one column

In [34]:
test_df5= test_df3.copy(deep=True)
test_df5

TITLE  ... BROWSE_NODE_ID
0                                                                                     pete cat bedtime blues doll   ...              0
1                                                                                new yorker  refrigerator magnet    ...              1
2      ultimate selfsufficiency handbook complete guide baking crafts gardening preserving harvest raising animals  ...              2
3                                                                      amway nutrilite kids chewable iron tablets   ...              3
4                                      teacher planner company   lesson academic teacher planner  leatherette pink  ...              4
...                                                                                                            ...  ...            ...
19995                                                                          dixcy scott womens skinny fit capri  ...           1058
19996   electroprime  sheilded cable management  universal keystone jacks multimedia panel  rackmount data cabinet  ...           2205
19997        electroprime  universal silver tone chrome plastic auto car wheel center hub cap cover dust againster  ...           2161
19998                                     electroprime  pcs  dia mesh car tweeter trumpet dustproof cover  toyota   ...          19998
19999                                                        reebok mens twist run multicolor shoes  ukindia   us   ...            701

[20000 rows x 5 columns]

In [35]:

test_df5['text']=test_df5['TITLE']+" "+test_df5['DESCRIPTION']+" "+test_df5['BULLET_POINTS']+" "+test_df5['BRAND']
test_df5

TITLE  ...                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          text
0                                                                                     pete cat bedtime blues doll   ...                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    pete cat bedtime blues doll  pete cat coolest popular cat town new pete cat bedtim blue doll merrymak rock stripe pj red slipper one sleepi cat readi cuddl measur tall safe age remov cloth surfac wash new pete cat bedtim blue plush doll base popular pete cat book jame dean super cuddli readi naptim bedtim safe age perfect age measur MerryMakers
1                                                                                new yorker  refrigerator magnet    ...                                                                                                                                                                                                                                                                                                                                                                                 new yorker  refrigerator magnet   new yorker handsom cello wrap hard magnet measur width height highlight one mani beauti new yorker cover full color cat tea cup new yorker cover artist gurbuz dogan eksioglu cat tea cup new yorker cover artist gurbuz dogan eksioglu handsom cello wrap hard magnet ideal home offic gift new yorker magazin lover highlight one mani beauti new yorker cover full color rigid magnet measur width height The New Yorker
2      ultimate selfsufficiency handbook complete guide baking crafts gardening preserving harvest raising animals  ...                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           ultimate selfsufficiency handbook complete guide baking crafts gardening preserving harvest raising animals  skyhors publish imusti
3                                                                      amway nutrilite kids chewable iron tablets   ...                                                                                                                                                                                                                                                              

In [36]:
test_df6=test_df5.copy(deep=True)
test_df6=test_df6.drop(columns=['TITLE','DESCRIPTION','BULLET_POINTS','BRAND'])
test_df6

BROWSE_NODE_ID                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          text
0                   0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    pete cat bedtime blues doll  pete cat coolest popular cat town new pete cat bedtim blue doll merrymak rock stripe pj red slipper one sleepi cat readi cuddl measur tall safe age remov cloth surfac wash new pete cat bedtim blue plush doll base popular pete cat book jame dean super cuddli readi naptim bedtim safe age perfect age measur MerryMakers
1                   1                                                                                                                                                                                                                                                                                                                                                                                 new yorker  refrigerator magnet   new yorker handsom cello wrap hard magnet measur width height highlight one mani beauti new yorker cover full color cat tea cup new yorker cover artist gurbuz dogan eksioglu cat tea cup new yorker cover artist gurbuz dogan eksioglu handsom cello wrap hard magnet ideal home offic gift new yorker magazin lover highlight one mani beauti new yorker cover full color rigid magnet measur width height The New Yorker
2                   2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           ultimate selfsufficiency handbook complete guide baking crafts gardening preserving harvest raising animals  skyhors publish imusti
3                   3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  

In [ ]:
test_df6.text.value_counts()

jack  jones mens slim fit cotton tshirt   cotton slim fit short sleev round neck hand wash Jack & Jones                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 

In [37]:
# Checking balance of dataset
grouped_df = test_df6.groupby('BROWSE_NODE_ID').count().values.flatten()
len(grouped_df)  # no of classes

3219

In [38]:
np.max(test_df6['text'].apply(lambda x: len(x.split())).values)
# Finding maximum length of text message

638

In [39]:
test_df7=test_df6.copy(deep=True)
test_df7

BROWSE_NODE_ID                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          text
0                   0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    pete cat bedtime blues doll  pete cat coolest popular cat town new pete cat bedtim blue doll merrymak rock stripe pj red slipper one sleepi cat readi cuddl measur tall safe age remov cloth surfac wash new pete cat bedtim blue plush doll base popular pete cat book jame dean super cuddli readi naptim bedtim safe age perfect age measur MerryMakers
1                   1                                                                                                                                                                                                                                                                                                                                                                                 new yorker  refrigerator magnet   new yorker handsom cello wrap hard magnet measur width height highlight one mani beauti new yorker cover full color cat tea cup new yorker cover artist gurbuz dogan eksioglu cat tea cup new yorker cover artist gurbuz dogan eksioglu handsom cello wrap hard magnet ideal home offic gift new yorker magazin lover highlight one mani beauti new yorker cover full color rigid magnet measur width height The New Yorker
2                   2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           ultimate selfsufficiency handbook complete guide baking crafts gardening preserving harvest raising animals  skyhors publish imusti
3                   3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  

In [40]:
#creating classes categorical wise
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
le.fit(test_df7['BROWSE_NODE_ID'])

test_df7['class_num'] = le.transform(test_df7['BROWSE_NODE_ID'])
test_df7

BROWSE_NODE_ID  ... class_num
0                   0  ...         0
1                   1  ...         1
2                   2  ...         2
3                   3  ...         3
4                   4  ...         4
...               ...  ...       ...
19995            1058  ...       682
19996            2205  ...      1114
19997            2161  ...      1100
19998           19998  ...      3218
19999             701  ...       472

[20000 rows x 3 columns]

In [41]:
test_df7['class_num'].max()

3218

In [42]:
test_df8= test_df7.copy(deep=True)
test_df8

BROWSE_NODE_ID  ... class_num
0                   0  ...         0
1                   1  ...         1
2                   2  ...         2
3                   3  ...         3
4                   4  ...         4
...               ...  ...       ...
19995            1058  ...       682
19996            2205  ...      1114
19997            2161  ...      1100
19998           19998  ...      3218
19999             701  ...       472

[20000 rows x 3 columns]

In [ ]:
# import seaborn as sns 
# import matplotlib.pyplot as plt 
# sns.set(style="whitegrid")

# f, ax = plt.subplots(figsize=(15, 10)) 
# sns.set_color_codes("pastel") 
# sns.barplot(y="reviews.rating", x="No of Users", data=test_df4['BRAND'].iloc[:20, :10],label="Score", color="pink")

# ax.legend(ncol=2, loc="upper left", frameon=True) 
# ax.set(xlabel="No of People",ylabel="Rating") 
# sns.despine(left=True, bottom=True) 
# plt.show()

In [44]:
# The maximum number of words to be used. (most frequent)
MAX_NB_WORDS = 80000
# Max number of words in each complaint.
MAX_SEQUENCE_LENGTH = 700
# This is fixed.
EMBEDDING_DIM = 100

tokenizer = Tokenizer(num_words=MAX_NB_WORDS, filters='!"#$%&()*+,-./:;<=>?@[\]^_`{|}~', lower=True)
tokenizer.fit_on_texts(test_df8['text'].values)
word_index = tokenizer.word_index
print('Found %s unique tokens.' % len(word_index))

Found 71398 unique tokens.


In [45]:
from keras.preprocessing.sequence import pad_sequences

In [46]:
X = tokenizer.texts_to_sequences(test_df8['text'].values)
X = pad_sequences(X, maxlen=MAX_SEQUENCE_LENGTH)
print('Shape of data tensor:', X.shape)

Shape of data tensor: (20000, 700)


Model making
- https://www.kaggle.com/namansood/nlp-xgboost-bert-glove-rnns-vectorizers 
- using this process of vectorization

In [47]:
x = test_df8['text']
y = test_df8['class_num']
len(x), len(y)

(20000, 20000)

In [48]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, y)
print(len(x_train), len(y_train))
print(len(x_test), len(y_test))

15000 15000
5000 5000


In [49]:
# First working with count vectorizer

from sklearn.feature_extraction.text import CountVectorizer

# instantiate the vectorizer
count = CountVectorizer()
count.fit(x)

x_train_num = count.transform(x_train)
x_test_num = count.transform(x_test)

In [48]:
# Example of a tuned model
count_tuned = CountVectorizer(stop_words='english', ngram_range=(1,2), min_df=0.1, max_df=0.7, max_features=4000)


In [49]:
# Working with TF-IDF now
from sklearn.feature_extraction.text import TfidfTransformer
# We are using transformer here
# If we use vectorizer, we can directly use the text
tfidf = TfidfTransformer()

tfidf.fit(x_train_num)
x_train_tfidf = tfidf.transform(x_train_num)

x_train_tfidf

<15000x69216 sparse matrix of type '<class 'numpy.float64'>'
	with 884445 stored elements in Compressed Sparse Row format>

In [51]:
text = test_df8['text']
c = test_df8['class_num']

In [52]:
# Calculating the total vocabulary
tk = Tokenizer()
tk.fit_on_texts(text)

vocab = len(tk.word_index)+1
vocab

71399

In [53]:
# Maximum length
max_len = np.max(test_df8['text'].apply(lambda x: len(x.split())).values)
max_len

638

In [54]:
def embedding(text):
    return tk.texts_to_sequences(text)

train_padded = pad_sequences(embedding(text), 638, padding='post')
train_padded

array([[12347,  1076, 34787, ...,     0,     0,     0],
       [   62,  8723,  7297, ...,     0,     0,     0],
       [ 5251, 34790, 34791, ...,     0,     0,     0],
       ...,
       [ 3203,  2270,   202, ...,     0,     0,     0],
       [ 3203,  1250,  2268, ...,     0,     0,     0],
       [ 3576,   136,  1525, ...,     0,     0,     0]], dtype=int32)

model creation and prediction

In [55]:
train_padded.shape

(20000, 638)

In [56]:
x_train, x_test, y_train, y_test = train_test_split(train_padded, c, test_size=0.2)

In [84]:
from keras.models import Sequential
from keras.layers import Dense,Flatten
from keras.backend import eval
from keras.optimizers import Adam
from keras.layers import LSTM
from keras.layers.embeddings import Embedding
from keras.layers import Dropout
from keras.layers.convolutional import Conv1D,MaxPooling1D
from keras.layers import SpatialDropout1D,Bidirectional,GlobalMaxPool1D,BatchNormalization

In [92]:
x = tokenizer.texts_to_sequences(test_df8['text'].values)
x = pad_sequences(x, maxlen=MAX_SEQUENCE_LENGTH)
print('Shape of data tensor:', x.shape)
y = pd.get_dummies(test_df8['BROWSE_NODE_ID']).values
print('Shape of label tensor:', y.shape)

Shape of data tensor: (20000, 700)
Shape of label tensor: (20000, 3219)


In [93]:
X_train, X_test, Y_train, Y_test = train_test_split(x,y, test_size = 0.10, random_state = 42)
print(X_train.shape,Y_train.shape)
print(X_test.shape,Y_test.shape)

(18000, 700) (18000, 3219)
(2000, 700) (2000, 3219)


In [107]:
# model = tf.keras.models.Sequential([
# model.add(Embedding(input_dim=embedding_matrix.shape[0], 
#                    output_dim=embedding_matrix.shape[1],
#                    weights=[embedding_matrix],
#                    input_length=max_len
#                    )
#          )
# model.add(Bidirectional(LSTM(max_len, return_sequences=True, recurrent_dropout=0.15)))
# model.add(GlobalMaxPool1D())
# model.add(BatchNormalization())
# model.add(Dropout(0.3))
# model.add(Dense(max_len, activation='relu'))
# model.add(Dropout(0.3))
# model.add(Dense(max_len, activation='relu'))
# model.add(Dropout(0.3))
# model.add(Dense(3218, activation='softmax'))

model = Sequential()
model.add(Embedding(MAX_NB_WORDS, EMBEDDING_DIM, input_length=x.shape[1]))
# model.add(BatchNormalization())
model.add(Dense(max_len, activation='relu'))
model.add(Dense(3219, activation='softmax'))
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

model.summary()

Model: "sequential_20"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_18 (Embedding)     (None, 700, 100)          8000000   
_________________________________________________________________
dense_24 (Dense)             (None, 700, 638)          64438     
_________________________________________________________________
dense_25 (Dense)             (None, 700, 3219)         2056941   
Total params: 10,121,379
Trainable params: 10,121,379
Non-trainable params: 0
_________________________________________________________________


In [108]:


epochs = 4
batch_size = 64

history = model.fit(x_train, y_train,batch_size = 64, epochs=5)


Epoch 1/5


ValueError: ignored

In [ ]:
#model - multiclass
# normal NN
import tensorflow as tf
import keras

model = tf.keras.models.Sequential([
    # # Note the input shape is the desired size of the image 150x150 with 3 bytes color
    # tf.keras.layers.Conv2D(16, (3,3), activation='relu', input_shape=(150, 150, 3)),
    # tf.keras.layers.MaxPooling2D(2,2),
    # tf.keras.layers.Conv2D(32, (3,3), activation='relu'),
    # tf.keras.layers.MaxPooling2D(2,2), 
    # tf.keras.layers.Conv2D(64, (3,3), activation='relu'), 
    # tf.keras.layers.MaxPooling2D(2,2),
    # Flatten the results to feed into a DNN
    tf.keras.layers.Flatten(), 
    # 512 neuron hidden layer
    tf.keras.layers.Dense(1024, activation='relu'), 
    # Only 1 output neuron. It will contain a value from 0-1 where 0 for 1 class ('cats') and 1 for the other ('dogs')
    tf.keras.layers.Dense(9919, activation=tf.nn.softmax)  
])

# model.add(Embedding(MAX_NB_WORDS, EMBEDDING_DIM, input_length=X.shape[1]))
# model.add(SpatialDropout1D(0.2))
# model.add(LSTM(100, dropout=0.2, recurrent_dropout=0.2))
# model.add(Dense(13, activation='softmax'))
# model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

# epochs = 5
# batch_size = 64

# history = model.fit(X_train, Y_train, epochs=epochs, batch_size=batch_size,validation_split=0.1,callbacks=[EarlyStopping(monitor='val_loss', patience=3, min_delta=0.0001)])


model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model.fit(training_images,  , epochs=5)
model.summary()
test_loss = model.evaluate(test_images, test_labels)


In [ ]:
sample_train['BULLET_POINTS']=sample_train['BULLET_POINTS'].apply(cleaning)
sample_train

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


TITLE  ... BROWSE_NODE_ID
0                                                                            pete cat bedtim blue doll  ...              0
1                                                                           new yorker refriger magnet  ...              1
2                  ultim selfsuffici handbook complet guid bake craft garden preserv harvest rais anim  ...              2
3                                                               amway nutrilit kid chewabl iron tablet  ...              3
4                                teacher planner compani lesson academ teacher planner leatherett pink  ...              4
...                                                                                                ...  ...            ...
19996       electroprim sheild cabl manag univers keyston jack multimedia panel rackmount data cabinet  ...           2205
19997  electroprim univers silver tone chrome plastic auto car wheel center hub cap cover dust against  ...           2161
19998                               electroprim pc dia mesh car tweeter trumpet dustproof cover toyota  ...          19998
19999                                                  reebok men twist run multicolor shoe ukindia us  ...            701
20000                                           qeboo girl cottonnet beauti dress blue year combo pack  ...           1687

[20001 rows x 5 columns]

In [ ]:
sample_train['BULLET_POINTS']=sample_train['BULLET_POINTS'].apply(sw_rem)
sample_train

In [ ]:
cross_val= sample_train.sample(frac=.2) #create validation set from training set
cross_val

In [ ]:
sample_train(['TITLE','DESCRIPTION','BRAND'])

In [ ]:
# The maximum number of words to be used. (most frequent)
MAX_NB_WORDS = 50000
# Max number of words in each complaint.
MAX_SEQUENCE_LENGTH = 250
# This is fixed.
EMBEDDING_DIM = 100
tokenizer = Tokenizer(num_words=MAX_NB_WORDS, filters='!"#$%&()*+,-./:;<=>?@[\]^_`{|}~', lower=True)
tokenizer.fit_on_texts(df['Consumer complaint narrative'].values)
word_index = tokenizer.word_index
print('Found %s unique tokens.' % len(word_index))